core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -504,17 +522,32 @@ public void run() {
 
   @Override
   public final void cancel(final Status reason) {
-    Substream noopSubstream = new Substream(0 /* previousAttempts doesn't matter here */);
-    noopSubstream.stream = new NoopClientStream();
-    Runnable runnable = commit(noopSubstream);
-
-    if (runnable != null) {
-      synchronized (lock) {
-        state = state.substreamDrained(noopSubstream);
+    // Handle everything the same way, e.g. isHedging vs not.
```

Comment:
This strategy plus attempting to brute-force flush in `drain()` (even non-conditionally as the code is now) is still apparently not enough to fix the leak and get rid of the explicit flush after writes. I think this is a reasonable attempt at what we discussed offline, am I missing anything obvious?